src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -445,54 +529,200 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            MethodSymbol? createSpan = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (createSpan is not null)
+            {
+                // CreateSpan was available. Use it.
+                // ldtoken <PrivateImplementationDetails>...
+                // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType>(fldHandle)
+                var field = _builder.module.GetFieldForData(data, alignment: (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Ldtoken);
+                _builder.EmitToken(field, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: 0);
+                EmitSymbolToken(createSpan.Construct(elementType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // We're dealing with a multi-byte primitive, and CreateSpan was not available.  Get a static field from PrivateImplementationDetails,
+            // and use it as a lazily-initialized cache for an array for this data:
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= RuntimeHelpers.InitializeArray(new int[Length], PrivateImplementationDetails.DataField));
+
+            var rosArrayCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Array, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosArrayCtor is null)
+            {
+                // The ReadOnlySpan<T>(T[] array) constructor we need is missing or something went wrong.
+                return false;
             }
+            Debug.Assert(!rosArrayCtor.HasUnsupportedMetadata);
 
+            // If we're dealing with an array of enums, we need to handle the possibility that the data blob
+            // is the same for multiple enums all with the same underlying type, or even with the underlying type
+            // itself. This is addressed by always caching an array for the underlying type, and then relying on
+            // arrays being covariant between the underlying type and the enum type, so that it's safe to do:
+            //     new ReadOnlySpan<EnumType>(arrayOfUnderlyingType);
+            // It's important to have a consistent type here, as otherwise the type of the caching field could
+            // end up changing non-deterministically based on which type for a given blob was encountered first.
+            if (elementType.IsEnumType())
+            {
+                arrayType = arrayType.WithElementType(TypeWithAnnotations.Create(arrayType.ElementType.EnumUnderlyingTypeOrSelf()));
```

Comment:
>TypeWithAnnotations.Create [](http://example.com/codeflow?start=54&length=26)

As we discussed offline, doing this also helps dropping custom modifiers, which potentially could be the source of similar non-deterministic behavior. I would call ```WithElementType``` unconditionally. That is unlikely to have any observable performance effect. So, trying to avoid creation of a new array symbol is probably not worth it.